packages/flutter/lib/src/widgets/resolution_awareness.dart

```diff
@@ -0,0 +1,127 @@
+/// ".meta". This metadata asset is a textual list of individual assets for
+/// that group, separated by newlines. Individual assets should be keyed with a
+/// prefix of "Nx/" where N is the nominal device pixel ratio for that asset.
+/// N does not need to be an integer.
+///
+/// For example, suppose an application wants to use a logical icon named
+/// "heart". This icon has representations at 1.0, 1.5, and 2.0 pixel ratios. The
+/// asset bundle should then contain the following assets:
+///
+/// heart.meta
+/// 1.0x/heart.png
+/// 1.5x/heart.png
+/// 2.0x/heart.png
+///
+/// On a device with a 1.0 device pixel ratio, the image chosen would be
+/// 1.0x/heart.png; on a device with a 1.3 device pixel ratio, the image chosen
+/// would be 1.5x/heart.png.
+///
+/// Common prefixes will be removed, so the following would be a valid bundle
+/// layout for an asset group named "icons/heart":
+///
+/// icons/heart.meta
+/// icons/1.0x/heart.png
+/// icons/1.5x/heart.png
+/// icons/2.0x/heart.png
+class ResolutionAwareAssetResolver extends CandidateAssetResolver {
+  ResolutionAwareAssetResolver({ AssetBundle bundle, this.context })
+    : super(bundle: bundle);
+
+  BuildContext context;
+
+  SplayTreeMap<double, String> _buildMapping(List<String> candidates) {
+    SplayTreeMap<double, String> result = new SplayTreeMap<double, String>();
+    RegExp exp = new RegExp(r"/?(\d+(\.\d*)?)x/");
+    for (int i = 0; i < candidates.length; i++) {
+      Match match = exp.firstMatch(candidates[i]);
+      if (match != null && match.groupCount > 0) {
+        double resolution = double.parse(match.group(1));
+        result[resolution] = candidates[i];
+      }
+    }
+    return result;
+  }
+
+  String _resolveFromMapping(BuildContext context, SplayTreeMap<double, String> mapping) {
+    MediaQueryData media = MediaQuery.of(context);
+    double value = media.devicePixelRatio;
+    BreakpointQualifier qualifier = new BreakpointQualifier(
+      candidates: mapping.values.toList(),
+      values: mapping.keys.toList(),
+      value: value
+    );
+    return qualifier.resolve();
+  }
+
+  String chooseCandidate(List<String> candidates) {
+    return _resolveFromMapping(context, _buildMapping(candidates));
+  }
+}
+
+/// Displays an image from an [AssetBundle] using the [ResolutionAwareAssetResolver]
+/// resolution strategy.
+class ResolutionAwareAssetImage extends AssetImage {
+  ResolutionAwareAssetImage({
+    Key key,
+    String name,
+    AssetBundle bundle,
+    double width,
+    double height,
+    ColorFilter colorFilter,
+    ImageFit fit,
+    FractionalOffset alignment,
+    ImageRepeat repeat: ImageRepeat.noRepeat,
+    Rect centerSlice
+  }) : super(
+    key: key,
+    name: name,
+    bundle: bundle,
+    width: width,
+    height: height,
+    colorFilter: colorFilter,
+    fit: fit,
+    alignment: alignment,
+    repeat: repeat,
+    centerSlice: centerSlice
+  );
+
```

Comment:
This seems like a lot of boilerplate just to change one constructor argument, but I didn't immediately see a better alternative.